File: sdks/python/endurance-tests/tests/test_pause_resume_churn.py
"""Endurance: one long-lived session, repeated pause/resume churn on a
recvonly track — no frames, no commands, no reconnect.

Isolates `Track.pause()`/`resume()` (`reactor_sdk/track.py`, backed by
`Reactor._pause_track`/`_resume_track`) as its own scenario rather than
folding it into `test_session_churn.py`'s already-mixed publish/frame/
command loop — a leak specific to the pause/resume path (e.g. the native
session-side pause state not fully clearing) should read as its own signal.

Manual-only for now (see ../README.md) — not wired into CI yet. Run with:

    ENDURANCE_DURATION_SECONDS=3600 mise run test:python:endurance-tests
"""

from __future__ import annotations

import sys
from pathlib import Path

from helpers import (
    ENDURANCE_DURATION_SECONDS,
    ResourceSampler,
    assert_always_zero,
    assert_never_grows,
    assert_no_sustained_growth,
    cpu_deltas,
)

sys.path.insert(0, str(Path(__file__).parent.parent))
from report import LiveReporter, MetricResult, finish_run, now_iso  # noqa: E402

import reactor_sdk
from reactor_sdk import Reactor

# The recvonly track reactor/echo always declares — same one
# test_lifecycle_churn.py/test_session_churn.py already subscribe to for
# on_frame, chosen here for the same reason: no need to invent a name.
TRACK_NAME = "main_video"


async def test_pause_resume_churn_has_no_sustained_growth(reactor: Reactor) -> None:
    sampler = ResourceSampler()
    live = LiveReporter("pause-resume-churn", ENDURANCE_DURATION_SECONDS)
    iteration = 0
    metrics: list[MetricResult] = []
    started_at = now_iso()
    # Resolves the track's direction from the session's already-declared
    # capabilities (see Reactor.track()) — a local, synchronous lookup, no
    # network round trip, so this is safe to do once before the loop rather
    # than on every iteration.
    track = reactor.track(TRACK_NAME)

    # Set on the first in-loop invariant violation below — same reasoning as
    # test_publish_churn.py's own flag.
    invariant_violated = False

    try:
        while not sampler.deadline_reached():
            await track.pause()
            await track.resume()

            # Exact, not trend-based: resume() above already returned, so the
            # session should report nothing paused on every single iteration —
            # `paused_tracks` is read fresh from the session (see its own
            # docstring), not cached, so a leftover entry here is a real
            # leftover pause state, not a stale local flag.
            paused = reactor.paused_tracks
            if paused:
                metrics.append(
                    MetricResult(
                        name="paused_tracks_after_resume",
                        start=0,
                        end=len(paused),
                        change=len(paused),
                        unit="count",
                        status="fail",
                        detail=(
                            f"{sorted(paused)} still reported paused after resume() "
                            f"on iteration {iteration}"
                        ),
                        threshold="always empty",
                        first_bad_cycle=iteration,
                    )
                )
                invariant_violated = True
                break

            sampler.sample(cycle=iteration)
            live.update(sampler.samples)
            iteration += 1

        if not invariant_violated:
            assert iteration >= 3, (
                f"only completed {iteration} iteration(s) — raise ENDURANCE_DURATION_SECONDS "
                "to get enough data for a trend"
            )

            # The fixture's one client is connected for the whole test, so the
            # baseline is 1, not 0 — same reasoning as the other two scenarios.
            metrics.append(assert_never_grows(sampler.samples, field="live_clients"))
            metrics.append(assert_always_zero(sampler.samples, field="orphaned_callbacks"))
            metrics.append(
                assert_no_sustained_growth(
                    [s.rss_bytes / 1e6 for s in sampler.samples],
                    name="rss",
                    unit="MB",
                    max_growth_ratio=0.15,
                    min_absolute_delta=5.0,
                )
            )
            metrics.append(
                assert_no_sustained_growth(
                    cpu_deltas(sampler.samples),
                    name="cpu_s_per_cycle",
                    unit="s",
                    max_growth_ratio=0.5,
                    min_absolute_delta=0.05,
                )
            )
            metrics.append(
                assert_no_sustained_growth(
                    [s.cpu_percent for s in sampler.samples],
                    name="cpu_percent",
                    unit="%",
                    max_growth_ratio=0.5,
                    min_absolute_delta=5.0,
                )
            )
            metrics.append(
                assert_no_sustained_growth(
                    [float(s.num_threads) for s in sampler.samples],
                    name="num_threads",
                    unit="count",
                    max_growth_ratio=0.15,
                    min_absolute_delta=4,
                    use_median=True,
                )
            )
            metrics.append(
                assert_no_sustained_growth(
                    [float(s.num_fds) for s in sampler.samples],
                    name="num_fds",
                    unit="count",
                    max_growth_ratio=0.15,
                    min_absolute_delta=3,
                )
            )
    finally:
        if sampler.samples:
            live.update(sampler.samples, force=True)

        result = finish_run(
            test_name="pause-resume-churn",
            sdk="Python",
            duration_s=ENDURANCE_DURATION_SECONDS,
            started_at=started_at,
            samples=sampler.samples,
            metrics=metrics,
            iterations=iteration,
            sdk_version=reactor_sdk.__version__,
            description=(
                "One long-lived session: repeated pause/resume on a recvonly "
                "track. No frames, no commands, no reconnect — isolates "
                "Track.pause()/resume() as its own signal."
            ),
        )

    if result.status == "FAIL":
        names = ", ".join(m.name for m in metrics if m.status == "fail")
        raise AssertionError(
            f"endurance test detected a failure in: {names} — see the report for details"
        )

File: sdks/cpp/endurance-tests/test_pause_resume_churn.cpp
// Endurance: one long-lived session, repeated pause/resume churn on a
// recvonly track — no frames, no commands, no reconnect. Mirrors
// sdks/python/endurance-tests/tests/test_pause_resume_churn.py.
//
// Isolates `Track::pause()`/`resume()` as its own scenario rather than
// folding it into test_session_churn.cpp's already-mixed publish/frame/
// command loop — a leak specific to the pause/resume path (e.g. the native
// session-side pause state not fully clearing) should read as its own
// signal. This is the scenario the Python suite's own first real CI run
// found a genuine, linear, no-plateau RSS climb on (+50% over 5 minutes)
// that publish-churn — same run, same process — did not show at all;
// folded into a broader mix that signal would have read as "RSS grew a bit,
// inconclusive" instead of naming pause/resume outright.
//
// Manual-only for now (see ../README.md). Run with:
//
//     ENDURANCE_DURATION_SECONDS=3600 mise run test:cpp:endurance-tests

#include <exception>
#include <optional>
#include <reactor/version.hpp>
#include <string>

#include <catch2/catch_test_macros.hpp>

#include "fixtures.hpp"
#include "helpers.hpp"
#include "report.hpp"
#include "trends.hpp"

namespace {
// The recvonly track reactor/echo always declares — the same one
// test_lifecycle_churn.cpp/test_session_churn.cpp already subscribe to for
// on_frame, chosen here for the same reason: no need to invent a name.
const char* const TRACK_NAME = "main_video";

const char* const DESCRIPTION =
    "One long-lived session: repeated pause/resume on a recvonly track. No "
    "frames, no commands, no reconnect — isolates Track::pause()/resume() as "
    "its own signal.";
}  // namespace

TEST_CASE("pause/resume churn has no sustained resource growth") {
  endurance::ResourceSampler sampler;
  endurance::LiveReporter live("pause-resume-churn", endurance::endurance_duration_seconds());
  int iteration = 0;
  std::exception_ptr pending;
  std::vector<endurance::MetricResult> metrics;
  const std::string started_at = endurance::now_iso();
  // Both constructed *inside* the try block below, not as plain locals
  // declared here: `ConnectedReactor`'s constructor is what actually
  // connects, and a connect failure there has to reach the catch block like
  // any other failure in this run, or finish_and_check() never runs and
  // this scenario silently writes no report at all — caught by actually
  // running this suite with no API key configured.
  std::optional<integration::ConnectedReactor> client_holder;
  std::optional<reactor::Track> track;

  // Same reasoning as test_publish_churn.cpp's own flag.
  bool invariant_violated = false;

  try {
    client_holder.emplace();
    auto& reactor = *client_holder;
    // Resolved once, up front: kind()/direction() are local, synchronous
    // lookups against the session's already-declared capabilities (see
    // Reactor::track()), so this is safe outside the loop rather than on
    // every iteration.
    track = reactor->track(TRACK_NAME);
    while (!sampler.deadline_reached()) {
      track->pause().get();
      track->resume().get();

      // Exact, not trend-based: resume() above already returned, so the
      // track should report not-paused on every single iteration — read
      // fresh from the session (see Track::paused()'s own docs), not
      // cached, so a leftover true here is a real leftover pause state.
      if (track->paused()) {
        endurance::MetricResult result;
        result.name = "track_paused_after_resume";
        result.start = 0;
        result.end = 1;
        result.change = 1;
        result.unit = "count";
        result.status = "fail";
        result.detail = std::string(TRACK_NAME) +
                        " still reports paused=true after resume() "
                        "on iteration " +
                        std::to_string(iteration);
        result.threshold = "always false";
        result.first_bad_cycle = iteration;
        metrics.push_back(result);
        invariant_violated = true;
        break;
      }

      sampler.sample(iteration);
      live.update(sampler.samples());
      ++iteration;
    }
  } catch (...) {
    pending = std::current_exception();
  }

  if (!invariant_violated && !pending) {
    if (iteration < 3) {
      pending = std::make_exception_ptr(std::runtime_error(
          "only completed " + std::to_string(iteration) +
          " iteration(s) — raise ENDURANCE_DURATION_SECONDS to get enough data for a trend"));
    } else {
      // The fixture's one client is connected for the whole test — same
      // reasoning as the other long-lived-session scenarios.
      //
      // `iteration < 3` above is a floor on completed iterations, not on
      // the sample count each metric actually needs — cpu_deltas() alone
      // drops one value relative to the raw samples, so a run landing at
      // exactly 3-6 iterations can still be short of what
      // assert_no_sustained_growth requires and throw here. Caught the
      // same way the run loop's own failures are, so a short run still
      // gets an ERROR report instead of an uncaught exception skipping
      // finish_and_check entirely.
      try {
        const auto shared = endurance::standard_resource_metrics(sampler.samples());
        metrics.insert(metrics.end(), shared.begin(), shared.end());
      } catch (...) {
        pending = std::current_exception();
      }
    }
  }

  endurance::FinishAndCheckArgs args;
  args.test_name = "pause-resume-churn";
  args.sdk = "C++";
  args.description = DESCRIPTION;
  args.sdk_version = std::string(reactor::version());
  args.duration_s = endurance::endurance_duration_seconds();
  args.started_at = started_at;
  args.metrics = metrics;
  args.iterations = iteration;
  args.run_error = pending;
  endurance::finish_and_check(sampler.samples(), live, args);

  if (pending) {
    std::rethrow_exception(pending);
  }
}
